packages/store/addon/-private/system/fetch-manager.ts

```diff
@@ -0,0 +1,554 @@
+import { default as RSVP, Promise } from 'rsvp';
+import { DEBUG } from '@glimmer/env';
+import { run as emberRunLoop } from '@ember/runloop';
+import Adapter from "@ember/test/adapter";
+import { AdapterCache } from "./adapter-cache";
+import { assert, warn, inspect } from '@ember/debug';
+import Snapshot from './snapshot';
+import { guardDestroyedStore, _guard, _bind, _objectIsAlive } from './store/common';
+import { normalizeResponseHelper } from './store/serializer-response';
+import { serializerForAdapter } from './store/serializers';
+import { InvalidError } from '@ember-data/adapter/error';
+import coerceId from './coerce-id';
+import Ember from 'ember';
+
+import {
+  _findHasMany,
+  _findBelongsTo,
+  _findAll,
+  _query,
+  _queryRecord,
+} from './store/finders';
+import RequestCache from "./request-cache";
+import { RecordData, RelationshipRecordData } from "./model/record-data";
+
+function payloadIsNotBlank(adapterPayload) {
+  if (Array.isArray(adapterPayload)) {
+    return true;
+  } else {
+    return Object.keys(adapterPayload || {}).length;
+  }
+}
+
+const emberRun = emberRunLoop.backburner;
+
+export interface PendingFetchItem {
+  identifier: RecordIdentifier,
+  queryRequest: Request,
+  resolver: RSVP.Deferred<any>,
+  options: any,
+  trace?: any
+}
+
+export interface PendingSaveItem {
+  resolver: RSVP.Deferred<any>,
+  snapshot: Snapshot,
+  identifier: RecordIdentifier,
+  options: any,
+  queryRequest: Request
+}
+
+export interface Operation {
+  op: string
+}
+
+export interface FindRecordQuery extends Operation {
+  op: 'findRecord'
+  recordIdentifier: RecordIdentifier
+  options: any
+}
+
+export interface SaveRecordMutation extends Operation {
+  op: 'saveRecord'
+  recordIdentifier: RecordIdentifier
+  options: any
+}
+
+export interface Request {
+  data: Operation[];
+  options?: any
+}
+
+export interface RequestState {
+  state: 'pending' | 'fulfilled' | 'rejected';
+  type: 'query' | 'mutation';
+  request: Request;
+  response?: Response;
+}
+
+export interface Response {
+  rawData: unknown;
+  data: unknown;
+}
+
+export default class FetchManager {
+  _pendingFetch: Map<string, PendingFetchItem[]>;
+  isDestroyed: boolean;
+  _adapterCache: AdapterCache;
+  _store: any;
+  requestCache: RequestCache;
+  _pendingSave: PendingSaveItem[];
+
+  constructor(adapterCache: AdapterCache, store: any) {
+    // used to keep track of all the find requests that need to be coalesced
+    this._pendingFetch = new Map();
+    this._adapterCache = adapterCache;
+    this._store = store;
+
+    this._pendingSave = [];
+    this.requestCache = new RequestCache();
+  }
+
+  /**
+    This method is called by `record.save`, and gets passed a
+    resolver for the promise that `record.save` returns.
+
+    It schedules saving to happen at the end of the run loop.
+
+    @method scheduleSave
+    @private
+    @param {InternalModel} internalModel
+    @param {Resolver} resolver
+    @param {Object} options
```

Comment:
remove the types from these JSDoc tags -- since the TS compiler can understand them, things can get strange with mixed messages (e.g., `identifier`)